client/meta_endpoint.go

```diff
@@ -0,0 +1,110 @@
+package client
+
+import (
+	"net/http"
+	"strings"
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+	"github.com/hashicorp/nomad/nomad/structs"
+	nstructs "github.com/hashicorp/nomad/nomad/structs"
+	"golang.org/x/exp/maps"
+)
+
+type NodeMeta struct {
+	c *Client
+}
+
+func newNodeMetaEndpoint(c *Client) *NodeMeta {
+	n := &NodeMeta{c: c}
+	return n
+}
+
+func (n *NodeMeta) Apply(args *structs.NodeMetaApplyRequest, reply *structs.NodeMetaResponse) error {
+	defer metrics.MeasureSince([]string{"client", "node_meta", "apply"}, time.Now())
+
+	// Check node write permissions
+	if aclObj, err := n.c.ResolveToken(args.AuthToken); err != nil {
+		return err
+	} else if aclObj != nil && !aclObj.AllowNodeWrite() {
+		return nstructs.ErrPermissionDenied
+	}
+
+	if len(args.Meta) == 0 {
+		return structs.NewErrRPCCoded(http.StatusBadRequest, "request missing required Meta object")
+	}
+	for k := range args.Meta {
+		if k == "" {
+			return structs.NewErrRPCCoded(http.StatusBadRequest, "metadata keys must not be empty")
+		}
+
+		// Actually do a modicum of validation
+		if strings.Contains(k, "*") {
+			return structs.NewErrRPCCoded(http.StatusBadRequest, "metadata keys cannot contain *")
+		}
```

Comment:
Do we want to be more aggressive about this constraint, considering these values will want to be interpolated with `${meta.<key>}`? 

On the one hand I don't think we've formalized the constraint in docs so existing metadata in HCL might not match any validation we do here. But on the other hand, it's easier to open up the space of characters allowed later than the other way around. On the other other hand, _maybe_ we can assume cluster administrators know what they want here?

Reply:
https://pkg.go.dev/github.com/hashicorp/hcl/v2@v2.16.0/hclsyntax#ValidIdentifier has a nice interface and the spec seems reasonable: https://github.com/hashicorp/hcl/blob/main/hclsyntax/spec.md#identifiers

Seems like a reasonable thing to try.

Reply:
Just a note that I used that hcl function, but I split on dots as Nomad itself uses dotted identifiers in node metadata (eg `connect.log_level`). Added a table test around the validation.